central/notifications/datastore/internal/writer/writer_impl.go

```diff
@@ -0,0 +1,139 @@
+package writer
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/notifications/datastore/internal/store"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/protoconv"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+var (
+	_ Writer = (*writerImpl)(nil)
+
+	rootNamespaceUUID = uuid.Nil
+)
+
+type writerImpl struct {
+	mutex sync.Mutex
+
+	buffer map[string]*storage.Notification
+	store  store.Store
+}
+
+func (c *writerImpl) read(id string) (*storage.Notification, bool) {
+	notification, found := c.buffer[id]
+	return notification, found
+}
+
+func (c *writerImpl) write(notification *storage.Notification) {
+	c.buffer[notification.GetId()] = notification
+}
+
+func (c *writerImpl) Upsert(ctx context.Context, notification *storage.Notification) error {
```

Comment:
After reviewing the datastore PR, and this one here, I realized we currently do not perform any SAC checks.

This means, that within `Upsert` and unauthorized client can inject notifications that later on will be written to the store when `Flush` is called by an authorized client. Additionally, an authorized client with `READ` permissions can also now `WRITE`.

We should definitely add this here. I feel like the correct place would be the writer interface compared to the datastore interface (within the datastore, all other places like `List, Get, Count` are covered by the generated code and the SAC checkers there).

Reply:
I noticed that when adding a SAC check on global `pkg/administration/events` level I ran into import cycles. Will have to see if we can resolve them.

Reply:
We've converged on this offline, for posterity: at the current time, the `Writer` will perform SAC checks.

The `Stream` currently does _not_ have any protection whatsoever, so theoretically any client may add events to the stream, subsequently also in the datastore without authorization.

I'll create a follow-up PR which will introduce a concept of authorization to the stream. Re-using SAC principles there (since we now _could_, `resources` have moved to `pkg/sac/resource`), it introduces cyclic imports due to log statements being within `pkg/sac` (when using the resource helper). We are probably going to use a similar concept to the declarative config one, where a context is enriched with a particular value being set, and afterwards within the stream we validate the given context.

I'll resolve this comment for now, and create the follow-up PR afterwards.